core/src/main/java/org/elasticsearch/common/cache/Cache.java

```diff
@@ -173,6 +174,10 @@ public Entry(K key, V value, long writeTime) {
         ReleasableLock writeLock = new ReleasableLock(segmentLock.writeLock());
 
         Map<K, Entry<K, V>> map = new HashMap<>();
+
+        // canonicalized mapping of keys
+        Map<K, WeakReference<K>> keys = new WeakHashMap<>();
```

Comment:
can we avoid the weak reference? would it be enough to also clear a key from this map whenever we clear a key from the main map?